synapse/storage/_base.py

```diff
@@ -42,11 +43,14 @@ def __init__(
         database: DatabasePool,
         db_conn: LoggingDatabaseConnection,
         hs: "HomeServer",
+        datastore: Optional["DataStore"] = None,
     ):
         self.hs = hs
         self._clock = hs.get_clock()
         self.database_engine = database.engine
         self.db_pool = database
+        # A reference back to the root datastore.
+        self.datastore = datastore
```

Comment:
This seems a little suspicious; seems like it might just give us another way to build circular relationships between stores

Reply:
> another way to build circular relationships between stores

I am not trying to prevent that with these changes, I see that as a bit of a non-goal honestly.

Reply:
My goal was to reduce the implicit inheritance relationship between the datastores, however. Hopefully that makes sense?

Reply:
I can see that moving from `self.my_foostore_method()` to `self.datastore.foo.my_foostoremethod()` is:

- explicit about the fact that it's a cross-store call
- greppable 
- probably better for typecheckers

My concern about the circular argument is that: I'm slightly worried that this makes it easier for any store to call out to any other store, which sounds like a kind of "soup" of stores that might be hard to conceptually reason about.

... that's a slippery slope argument though. I think this is probably a change for the better

Reply:
> I can see that moving from `self.my_foostore_method()` to `self.datastore.foo.my_foostoremethod()` is:
> 
>  * explicit about the fact that it's a cross-store call
>  * greppable
>  * probably better for typecheckers

This is very much what I was hoping to solve!

> My concern about the circular argument is that: I'm slightly worried that this makes it easier for any store to call out to any other store, which sounds like a kind of "soup" of stores that might be hard to conceptually reason about.
> 
> ... that's a slippery slope argument though. I think this is probably a change for the better

It agree that it *might* make it slightly easier to call into other datastores. I think we could (in the future) block this by not giving each datastore a reference to "main" datastore.

Reply:
Actually, the relations stuff doesn't need access to that, so maybe we nuke it for now and try to clean it up as we go? WDYT?